### ISSUE TITLE
Add PolygonPatch object

### ISSUE THREAD

**Issue by story645** [matplotlib/matplotlib#16662]
# Problem
Currently svgs/shapely type polygon objects are a bit hard to plot directly in MPL 'cause you have to manually construct the path from them.  and construct a Matplotlib patch out of it. It'd be helpful for downstream libraries if there was code in MPL that directly plotted these objects (probably especially ducktyped against shapely since that's a standard.)

Here's a sample implementation by @jorisvandenbossche from https://github.com/geopandas/geopandas/issues/1039
```python
from matplotlib.patches import PathPatch
from matplotlib.path import Path
from numpy import asarray, concatenate, ones


def PolygonPatch(polygon, **kwargs):

    def coding(ob):
        # The codes will be all "LINETO" commands, except for "MOVETO"s at the
        # beginning of each subpath
        n = len(getattr(ob, 'coords', None) or ob)
        vals = np.ones(n, dtype=Path.code_type) * Path.LINETO
        vals[0] = Path.MOVETO
        return vals

    vertices = np.concatenate(
        [np.asarray(polygon.exterior)[:, :2]] +
        [np.asarray(r)[:, :2] for r in polygon.interiors])
    codes = np.concatenate(
        [coding(polygon.exterior)] + [coding(r) for r in polygon.interiors])

    return PathPatch(Path(vertices, codes), **kwargs)
```
@tacaswell proposed that this should directly be a subclass on the PathPatch object, becoming a new artist.

**Comment by ImportanceOfBeingErnest** [comment]
This brings a couple of assumptions about the `polygon` object into matplotlib code base. Is there a specific reason to not rely on `descartes` in such case?

**Comment by tacaswell** [comment]
True, but it does not bring in a dependency and the required structure in pretty simple.  The point is that they are relying on decartes for _just_ these 20 lines.

**Comment by anntzer** [comment]
I agree with @ImportanceOfBeingErnest that bringing in assumptions about shapely objects (namely, `.exterior` and `.interior` and `.coords` their layout) looks not nice -- just see the problems we have with handling pandas dataframes.

Also the implementation looks overly complicated, I think it's really just (modulo `.coords`, which I don't know what it's about) `PathPatch(Path.make_compound_path(Path(polygon.exterior), Path(polygon.interior)), **kwargs)`?  (you don't need to set the codes, and it's already a 2d array?

If anything, this *possibly* suggests that `make_compound_path` could get a shorter spelling, like `path1.followed_by(path2)` or (??) `path1 + path2`.

**Comment by tacaswell** [comment]
The main problem with Pandas is that we try to duck-type it as array like, this seems much easier to manage.

**Comment by anntzer** [comment]
Still, I have no idea what .exterior/.interior means, or when does the .coords attribute exists or not, and I'm not sure it's something I "want" to know...

**Comment by jklymak** [comment]
I'm not thrilled w/ the idea of supporting a particular 3rd downstream data type here either.  Sure, 20 lines of code here, 20 lines of code there, but there are *lots* of downstream packages.  

I think there is an argument to be made that we need a higher level `patch` object, I'd buy that.  `ax.simple_patch(xy, **kwargs)` seems better than `ax.add_patch(PathPatch(Path(vertices, codes), **kwargs))`.  Then downstream libraries could unpack their object into vertices without having to also come up with something as low level as the `codes` argument.

**Comment by anntzer** [comment]
Again, as stated above, you even don't actually need to set codes in this case.  So it's pretty close to just chaining 3 functions as is...

**Comment by jklymak** [comment]
Sure but chaining three obscure functions is fine for downstream packages, but obscure for users.

**Comment by anntzer** [comment]
As stated above, we should make them easier to use individually.
